Web Inspector: console's code completion should be case-insensitive

the-chenergy opened the pull request:
#### 14ea1c16323321db0dd237aad3aa22a9e48f9a54
<pre>
Web Inspector: console&apos;s code completion should be case-insensitive
<a href="https://rdar.apple.com/124544458">rdar://124544458</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=270925">https://bugs.webkit.org/show_bug.cgi?id=270925</a>

Reviewed by Devin Rousso.

Convert both strings into lowercase when trying to match prefixes for
function, variable, or property names.

Also add an experimental setting item for this feature enhancement.

* Source/WebInspectorUI/UserInterface/Controllers/CodeMirrorCompletionController.js:
(WI.CodeMirrorCompletionController.prototype._applyCompletionHint.update):
(WI.CodeMirrorCompletionController.prototype._applyCompletionHint):
   - Since case-insensitive matching may be used, only show the
     watermark text on the prompt if the typed text exactly,
     case-sensitively matches the prefix.

* Source/WebInspectorUI/UserInterface/Controllers/CodeMirrorCompletionController.js:
(WI.CodeMirrorCompletionController.prototype._generateJavaScriptCompletions.):
(WI.CodeMirrorCompletionController.prototype._generateJavaScriptCompletions):
* Source/WebInspectorUI/UserInterface/Controllers/JavaScriptRuntimeCompletionProvider.js:
(WI.JavaScriptRuntimeCompletionProvider.prototype.completionControllerCompletionsNeeded.receivedPropertyNames):
(WI.JavaScriptRuntimeCompletionProvider.prototype.completionControllerCompletionsNeeded):
(WI.JavaScriptRuntimeCompletionProvider.prototype.completionControllerCompletionsNeeded.receivedPropertyNames.compare): Deleted.
   - Optionally perform case-insensitive prefix matching when filtering
     the code completion suggestions.

* Source/WebInspectorUI/UserInterface/Base/Setting.js:
* Source/WebInspectorUI/UserInterface/Views/SettingsTabContentView.js:
* Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:
   - Add the experimental setting item for this change in the Settings
     tab&apos;s Experimental sub-tab.

* LayoutTests/inspector/console/js-completions-expected.txt:
* LayoutTests/inspector/console/js-completions.html:
   - Add a test case for the code completions&apos; ordering since now it
     gets slightly more involved with case-insensitive matching.

Canonical link: <a href="https://commits.webkit.org/279195@main">https://commits.webkit.org/279195@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4bc910057bc9a885adc431fdf347f4b7401e9230

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [  ~~🧪 style~~](https://ews-build.webkit.org/#/builders/38/builds/52614 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/48/builds/31949 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/55/builds/5044 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/55889 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/59/builds/3337 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") 
| | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/49/builds/38531 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/3038 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/55889 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 wincairo-tests~~](https://ews-build.webkit.org/#/builders/59/builds/3337 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") 
| [  ~~🧪 webkitperl~~](https://ews-build.webkit.org/#/builders/11/builds/54712 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/49/builds/38531 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/55/builds/5044 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/55889 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/49/builds/38531 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/55/builds/5044 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/1496 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/49/builds/38531 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/55/builds/5044 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/57484 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/27750 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/3038 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/57484 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/28975 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/55/builds/5044 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/57484 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11527 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/29889 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/28725 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/CodeMirrorCompletionController.js:
```diff
@@ -683,7 +684,7 @@ WI.CodeMirrorCompletionController = class CodeMirrorCompletionController extends
                 if (declaringVariable && !allowedKeywordsWhenDeclaringVariable.has(keyword))
                     continue;
-                if (!keyword.startsWith(prefix))
+                if (!keyword.toLowerCase().startsWith(prefix.toLowerCase()))
```
can we move `prefix.toLowerCase()` outside so that we dont keep having to call it?

unless `prefix` is modified somewhere?

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/CodeMirrorCompletionController.js:
```diff
@@ -699,7 +700,7 @@ WI.CodeMirrorCompletionController = class CodeMirrorCompletionController extends
                         continue;
 
-                    if (variable.name.startsWith(prefix) && !matchingWords.includes(variable.name))
+                    if (variable.name.toLowerCase().startsWith(prefix.toLowerCase()) && !matchingWords.includes(variable.name))
```
ditto

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/JavaScriptRuntimeCompletionProvider.js:
```diff
@@ -361,7 +361,7 @@ WI.JavaScriptRuntimeCompletionProvider = class JavaScriptRuntimeCompletionProvid
                 }
 
-                if (!property.startsWith(prefix) || knownCompletions.has(property))
+                if (!property.toLowerCase().startsWith(prefix.toLowerCase()) || knownCompletions.has(property))
```
ditto

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/JavaScriptRuntimeCompletionProvider.js:
```diff
@@ -383,6 +383,13 @@ WI.JavaScriptRuntimeCompletionProvider = class JavaScriptRuntimeCompletionProvid
+                if (aCaseSensitiveMatch && !bCaseSensitiveMatch)
+                    return -1;
+                if (!aCaseSensitiveMatch && bCaseSensitiveMatch)
+                    return 1;
```
woah, is `String.prototype.localeCompare` really case insensitive?  i feel like we should really be fixing that to care about case sensitivity instead of just changing it here

the-chenergy replied:
This is where I preserved the cases the user has typed. Without these lines, upper case would always come first. (`localeCompare` is case sensitive indeed)

Do you see a simpler way of doing this, BTW?

dcrousso replied:
oh i see.  this isn't so much about whether it's case sensitive, but about whether the user typed uppercase vs lowercase and preference based on that.  in that case this seems fine.  unfortunate that we're having to do _another_ two string comparisons each time we compare two strings, but i suppose this is the only way

the-chenergy replied:
> unfortunate that we're having to do another two string comparisons each time we compare two strings, but i suppose this is the only way

If we wanna save as much work as possible about comparing the strings while sorting, we could consider pre-computing the sorting criteria and make a tuple of data for each string, like this:
```js
// completions = ["string1", "string2", ...];

let completionSortCriteria = {};
for (let completion of completions) {
    let numericValue = isNaN(completion) ? Infinity : +completion;
    let isRareProperty = completion.startsWith("__") && completion.endsWith("__");
    let isNotExactMatch = !completion.startsWith(prefix);
    completionSortCriteria[completion] = [numericValue, isRareProperty, isNotExactMatch];
}

function compare(s, t)
{
    let a = completionSortCriteria[s];
    let b = completionSortCriteria[t];
    return a[0] - b[0] || a[1] - b[1] || a[2] - b[2] || s.extendedLocaleCompare(t);
}

completions.sort(compare);
```

Let me know if you think this would make a good fix.

dcrousso replied:
yeah i like that!

id probably use a `Map` instead of `{}`, and have the value be an object `{numericValue, isRareProperty, hasMatchingPrefix}`, but otherwise i am a fan :)

also, should this (perhaps preemptively) use `WI.CSSCompletions.getCompletionText` (tho we'd likely wanna move it to a more generic object)?

dcrousso reviewed (commented):
this definitely feels like something that should be rolled out with an on-by-default `WI.ExperimentalSetting`, since it could be really funky for developers that do rely on case sensitivity (e.g. `window.navigator` vs `window.Navigator`)

perhaps we could even have a followup `WI.Setting` that allows the developer to indicate whether lowercase or uppercase results should be weighted more heavily (e.g. maybe most/all of their functions are uppercased like some code style guides suggest)

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/JavaScriptRuntimeCompletionProvider.js:
```diff
@@ -372,30 +374,36 @@ WI.JavaScriptRuntimeCompletionProvider = class JavaScriptRuntimeCompletionProvid
-                return a.extendedLocaleCompare(b);
+                let a = completionSortCriteria.get(s);
+                let b = completionSortCriteria.get(t);
+                return (
```
Style: the `(` `)` are unnecessary
```js
return a.numericValue - b.numericValue
    || a.isRareProperty - b.isRareProperty
    || b.startsWithPrefix - a.startsWithPrefix
    || s.extendedLocaleCompare(t);
```

also should it be `a.startsWithPrefix - b.startsWithPrefix` instead?

the-chenergy replied:
We want the ones that `startsWithPrefix` to come first, which are the "greater" values to come first (`true` > `false`), which is the reverse of the typical "less coming first".

I can see that this code is hacky and potentially hard to understand on first glance, but it is _much_ shorter than its fully-expanded form (like in the old code) so I thought it was worth shortening it this way. I can expand it if you think that's still better.

dcrousso replied:
oh i also prefer the shorter form :)

it just might be nice to re-add some of the comments for clarity
```js
return a.numericValue - b.numericValue // Try to sort in numerical order first.
    || a.isRareProperty - b.isRareProperty // Sort __defineGetter__, __lookupGetter__, and friends last.
    || b.startsWithPrefix - a.startsWithPrefix // Prefer completions that directly match the typed text. 
    || s.extendedLocaleCompare(t); // Not numbers, sort as strings.
```

btw can we write tests for this?  perhaps we could add another test case to `LayoutTests/inspector/console/js-completions.html`?

the-chenergy commented on Source/WebInspectorUI/UserInterface/Controllers/JavaScriptRuntimeCompletionProvider.js:
```diff
@@ -372,30 +374,34 @@ WI.JavaScriptRuntimeCompletionProvider = class JavaScriptRuntimeCompletionProvid
+                let b = completionSortCriteria.get(t);
+                return a.numericValue - b.numericValue // Order completions that are numbers to come first, with lower values coming first.
+                    || a.isRareProperty - b.isRareProperty // Order __defineGetter__, __lookupGetter__, and friends to come last.
+                    || b.startsWithPrefix - a.startsWithPrefix // Order completions that directly match the typed text to come first.
```
Wait, shouldn't `startsWithPrefix` be given the highest weight so this check, `b.startsWithPrefix - a.startsWithPrefix`, should be done first? 🤔

dcrousso replied:
does it actually matter?  aren't all three of these checks (mostly) mutually exclusive?

regardless, i suppose `b.startsWithPrefix - a.startsWithPrefix` could go first just in case the developer is explicitly trying to fetch a rare property (e.g. `__proto__`) or a number property

the-chenergy replied:
Hmm right, this might not matter most of the time.

I suppose if in the future we end up adding fuzzy search to JS completion then this _might_ start mattering, where we will need to make sure `startsWithPrefix` gets checked first. Until then I think it's a good idea anyway since in principle that's what we want the sorting to do.

dcrousso commented on LayoutTests/inspector/console/js-completions.html:
```diff
@@ -52,6 +52,50 @@
+            };
+
+            await verifyCompletionsOrdering("");
+            await verifyCompletionsOrdering("doc");
```
can we try a few more?
```js
await verifyCompletionsOrdering("__pro");
await verifyCompletionsOrdering("1");
```
and maybe add a `window["123"] = "foo"` as well as a `window["__pro_test"] = "normal"` and `window["__pro_test__"] = "rare"`?

dcrousso commented on LayoutTests/inspector/console/js-completions-expected.txt:
```diff
@@ -13,3 +13,5 @@ PASS: Completions should still contain myVariable before cache is cleared.
 PASS: Completions should not contain myVariable after it's deleted by code run in console.
 
+-- Running test case: console.jsCompletions.completionOrdering
+
```
we should at least have _something_ be output from the test as otherwise if this becomes flaky it's much harder to understand where in the tests things are going wrong

i.e. i'd err on the side of over-logging (even if it's repetitive) as that there's at least some record of what's going on

the-chenergy commented on LayoutTests/inspector/console/js-completions.html:
```diff
@@ -5,28 +5,28 @@
+            },
+        };
+
+        const defaultCompletions = options.defaultCompletions || [];
```
These `const` should be `let` given how they're initialized now.

the-chenergy commented on LayoutTests/inspector/console/js-completions.html:
```diff
@@ -52,6 +52,50 @@
+        name: "console.jsCompletions.completionOrdering",
+        description: "Test that the JavaScript completions order its items in the expected way.",
+        async test() {
+            const verifyCompletionsOrdering = async (prefix) => {
```
Completions -> Completion perhaps

the-chenergy commented on Source/WebInspectorUI/UserInterface/Controllers/JavaScriptRuntimeCompletionProvider.js:
```diff
@@ -372,30 +374,35 @@ WI.JavaScriptRuntimeCompletionProvider = class JavaScriptRuntimeCompletionProvid
+                    startsWithPrefix: completion.startsWith(prefix),
+                    numericValue: isFinite(completion) ? +completion : Infinity,
+                    isRareProperty: (completion.startsWith("__") && completion.endsWith("__"))
+                        || (completion.startsWith("\"__") && completion.endsWith("__\"")),
```
It barely matters to check the `"__rareProperty__"` case with the surrounding quotes, but since I added a test case whose expected results do include them with quotes, I thought I'd keep it consistent anyway so they don't look surprising in the expected test results.

dcrousso replied:
do we even generate completions if the developer types quotes?

regardless, im not sure this is actually necessary, as if the property has quotes in it then it's not really a rare property (i.e. the engine will never add quotes to things like `__proto__`)

id probably leave this as it was.  we can always add this behavior in a followup if we feel it's useful :)

dcrousso commented on LayoutTests/inspector/console/js-completions.html:
```diff
@@ -52,6 +52,27 @@
+        description: "Test that the JavaScript completions are ordered correctly.",
+        async test() {
+            await InspectorTest.evaluateInPage(`myObject = {
+                1: 42, 2: 42, 10: 42, 20: 42, aa: 42, ab: 42, Aa: 42, Ab: 42, __aa__: 42, __ab__: 42
```
NIT: rather than do this as part of the test, id just declare this at the top of the test file
```js
window.myObject = {
    1: 42,
    ...
    __ab__: 42,
};
```

the-chenergy replied:
Sure, but why on top of the test file? I like that it's in this test function and at least look isolated from the other test cases.

dcrousso replied:
the main reason being that then other test cases can use it too if they want

also, having a test case "pollute" the global object can be harder to maintain if subsequent tests also want to do the same (e.g. what if we added another test case that also wanted to create a custom global `myObject`?)

FWIW we do this a lot with tests involving DOM nodes or CSS stuff and almost always prefer just having it exist before the test even runs (i.e. part of the file instead of evaluated after)

unless you're specifically trying to observe the act of "polluting" the global object id suggest avoiding having that as part of the test case :)

dcrousso commented on LayoutTests/inspector/console/js-completions.html:
```diff
@@ -52,6 +52,27 @@
+            };`);
+            WI.javaScriptRuntimeCompletionProvider.clearCachedPropertyNames();
+
+            InspectorTest.json(await generateJSCompletions({baseString: "myObject.", prefix: ""}));
```
NIT: it would be nice to maybe at least log something like "Generating completions for `myObject.`..." before each of these, as that way there's at least some context in the output file (i.e. if this test becomes flaky then it's potentially possible to just look at the output to see what went wrong instead of also having to look at this file too)

dcrousso commented on LayoutTests/inspector/console/js-completions.html:
```diff
@@ -52,6 +52,27 @@
+
+            InspectorTest.json(await generateJSCompletions({baseString: "myObject.", prefix: ""}));
+            InspectorTest.json(await generateJSCompletions({baseString: "myObject.", prefix: "a"}));
+            InspectorTest.json(await generateJSCompletions({baseString: "myObject.", prefix: "A"}));
```
should we do a few of these same tests again with `WI.settings.experimentalShowCaseSensitiveAutocomplete` flipped?

the-chenergy replied:
Good idea from the test's standpoint, but is this necessary if we plan to eventually remove that setting and always default to it being OFF anyway?

dcrousso replied:
what makes you say for sure that we'll end up removing it?  we've had settings for to control case-sensitive search for years.  i personally would strongly prefer keeping it case-sensitive, but i realize that's my preference so i'm fine with adding a way for others to have a different experience

also, if we do end up removing it, until we do we should test it :)

dcrousso commented on LayoutTests/inspector/console/js-completions.html:
```diff
@@ -52,6 +52,27 @@
+        name: "console.jsCompletions.completionOrdering",
+        description: "Test that the JavaScript completions are ordered correctly.",
+        async test() {
+            await InspectorTest.evaluateInPage(`myObject = {
```
if you truly want to have full control over what's in the object you could do `Object.create(null)` as that'll prevent any of the default stuff :)

dcrousso reviewed (approved):
nice work! 🤩

dcrousso commented on LayoutTests/inspector/console/js-completions.html:
```diff
@@ -52,6 +65,44 @@
+            await outputCompletions(({baseString: "objectWithMixedPropertyKinds[", prefix: "\"A"}));
+            await outputCompletions(({baseString: "objectWithMixedPropertyKinds[", prefix: "\"_"}));
+
+            WI.Setting.reset();
```
NIT: we should really just be calling `WI.settings.experimentalShowCaseSensitiveAutocomplete.reset()` since that's the only thing we're modifying in this test, and `WI.Setting.reset()` also wont modify the cached `_value` stored in `WI.settings.experimentalShowCaseSensitiveAutocomplete`

the-chenergy committed:
Web Inspector: console's code completion should be case-insensitive
rdar://124544458
https://bugs.webkit.org/show_bug.cgi?id=270925

Reviewed by Devin Rousso.

Convert both strings into lowercase when trying to match prefixes for
function, variable, or property names.

Also add an experimental setting item for this feature enhancement.

* Source/WebInspectorUI/UserInterface/Controllers/CodeMirrorCompletionController.js:
(WI.CodeMirrorCompletionController.prototype._applyCompletionHint.update):
(WI.CodeMirrorCompletionController.prototype._applyCompletionHint):
   - Since case-insensitive matching may be used, only show the
     watermark text on the prompt if the typed text exactly,
     case-sensitively matches the prefix.

* Source/WebInspectorUI/UserInterface/Controllers/CodeMirrorCompletionController.js:
(WI.CodeMirrorCompletionController.prototype._generateJavaScriptCompletions.):
(WI.CodeMirrorCompletionController.prototype._generateJavaScriptCompletions):
* Source/WebInspectorUI/UserInterface/Controllers/JavaScriptRuntimeCompletionProvider.js:
(WI.JavaScriptRuntimeCompletionProvider.prototype.completionControllerCompletionsNeeded.receivedPropertyNames):
(WI.JavaScriptRuntimeCompletionProvider.prototype.completionControllerCompletionsNeeded):
(WI.JavaScriptRuntimeCompletionProvider.prototype.completionControllerCompletionsNeeded.receivedPropertyNames.compare): Deleted.
   - Optionally perform case-insensitive prefix matching when filtering
     the code completion suggestions.

* Source/WebInspectorUI/UserInterface/Base/Setting.js:
* Source/WebInspectorUI/UserInterface/Views/SettingsTabContentView.js:
* Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:
   - Add the experimental setting item for this change in the Settings
     tab's Experimental sub-tab.

* LayoutTests/inspector/console/js-completions-expected.txt:
* LayoutTests/inspector/console/js-completions.html:
   - Add a test case for the code completions' ordering since now it
     gets slightly more involved with case-insensitive matching.

Canonical link: https://commits.webkit.org/279195@main
